README
Please have a look at the README in the doc/ directory...


HISTORY
joyd 0.2.2 2000-10-11
~~~~~~~~~~~~~~~~~~~~~
- a call without any command line arguments now successfully
  reads the compiled in default configuration file
- fixed some typos in the documentation



joyd 0.2.1 2000-05-25
~~~~~~~~~~~~~~~~~~~~~
 - fixed permissions in tar.gz-archive (everything was 0777)
 - config.stdout renamed to config.std_out due to an error in 
   /usr/include/stdio.h in older libc6 systems



joyd 0.2.0 2000-04-22
~~~~~~~~~~~~~~~~~~~~~
 - support of GNU autoconf/configure
 - documentation is now by default installed under /usr/local/doc/joyd
 - added a configuration example for control sequences (joydrc.sequence)
 - "joyd --version" will now also print version number (like "joyd -v")



joyd 0.0.7 2000-04-13
~~~~~~~~~~~~~~~~~~~~~
 - removed type cast of fgetc(), casting EOF to char instead
 - changed some C++ style comment lines to C style comments



joyd 0.0.6 2000-04-06
~~~~~~~~~~~~~~~~~~~~~
 - merged joyreadbutton.c and joyreadaxis.c into joyread.c
 - type cast of fgetc() to signed char because of compiler error
   on PPC systems
 - program version displayed on startup still was "0.0.3"
 - a new entry to the joystick configuration file (CONFIG)



joyd 0.0.5 2000-03-19
~~~~~~~~~~~~~~~~~~~~~
 - fixed a problem with joysticks that report buttons as axes with
   values of 1 and -1
 - added joyreadbutton and joyreadaxis to the joyd package
 - first two entries in the joystick configuration file (CONFIG)



joyd 0.0.4 2000-02-09
~~~~~~~~~~~~~~~~~~~~~
 - again some minor changes to the documentation, i.e. the default
   values are now listed in the joydrc(5) manpage



joyd 0.0.3 2000-01-17
~~~~~~~~~~~~~~~~~~~~~
 - calibration values are all wrong: the range is from -32768 to
   32767, not from -16384 to 16383
 - minor changes in documentation
 - added version information to the header files



joyd 0.0.2 2000-01-10
~~~~~~~~~~~~~~~~~~~~~
 - the global variable config was not declared correctly 
 - manual pages were added
 - "joyd -v" now prints the version number
 - config.c and config.o have been renamed to options.c and options.h
   because config.h files are generated by autoconfigure
 - the ActionLoop() function got split up into smaller parts



joyd 0.0.1 2000-01-09
~~~~~~~~~~~~~~~~~~~~~
first version
 - added signal handlers
 - added syslog routines
 - added background mode


joyd.c
/*
 *    joyd.c
 *
 *    this is the main file of:
 *
 *    joyd 0.0.7   ---   The Joystick Daemon
 *
 *    2000 (C) by Christian Garbs <mitch@uni.de>
 */

/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

/*
 * You can execute commands by using your joystick, joypad or whatever
 * is supported by the Linux Joystick Driver. This allows you to
 * reboot your computer when your keyboard is hung or you can even use
 * your system without any keyboard at all. 
 *
 * I use one of my computers (no screen, no keyboard) for listening to
 * mp3-files - with this program I can stop/start/skip songs, change
 * the volume, shutdown the computer and much more just by using an
 * joypad (in fact, it's an original SNES joypad).
 *
 * If you use this and find it useful, please let me know!
 * If you find a bug or have any suggestions or questions: tell me!
 *
 * My mail adress is:    <mitch@uni.de>
 * My homepage is here:  http://www.h.shuttle.de/mitch/
 *
 *
 *                                    Now enjoy this.
 *                                    Christian
 */

/*
 * This program is based on jstest.c Version 1.2 by Vojtech Pavlik,
 * which "is also intended to serve as an example implementation for
 * those who wish to learn how to write their own joystick using
 * applications."
 *
 * That's just what I did. That's GPL. Thanks!
 */


/*****************************************************************************
 *                                                                           *
 *     "Even the little things make me happy. It's that... that ordinary     *
 *      life I want back... I want back."                (Tsukino Usagi)     *
 *                                                                           *
 *****************************************************************************/


/*     These are just the changes to this file.
 *     See the HISTORY file for changes on the whole package.
 */


/*  joyd 0.0.1 2000-01-09
 *
 *  first version
 *  - added signal handlers
 *  - added syslog routines
 *  - added background mode
 */

/*  joyd 0.0.2 2000-01-10
 *
 *  - the global variable config is now declared here and declared
 *    external in joyd.h
 *  - "joyd -v" prints the version number
 */

/*  joyd 0.0.3 2000-01-17
 *
 *  - changed the JOYD_VERSION macro to PROGRAM_VERSION
 */

/*  joyd 0.0.4 2000-02-09
 *  joyd 0.0.5 2000-03-19
 *  joyd 0.0.6 2000-04-06
 *  joyd 0.0.7 2000-04-13
 *
 *  (no changes)
 */

/*
 *  2do:
 *  ? supply Sys V start/stop script (for /etc/rc.d/)
 *  ? implement shift_axes like shift_keys
 *  ? autoconf support
 */

/*****************************************************************************
 * now goes it loose...                                                      *
 *****************************************************************************/


/*
 * system include files
 */

#include <stdio.h>
#include <stdlib.h>

/*
 * my headers and definitions
 */

#include "joyd.h"
#include "daemon.h"
#include "joystick.h"
#include "log.h"
#include "options.h"
#include "signal.h"
#include "string.h"

/*
 * the global variable
 */

TCONFIG config;

/*
 * now to the functions
 */



int main (int argc, char **argv)
/* simple main program */
{
	if ((argc > 1) && (strcmp(argv[1],PRINT_VERSION) == 0)) {
		printf(PROGRAM_VERSION "\n");
		exit(0);
	};

	SetDefaultValues();

	if (config.debug > 1) {
		Print(stdout,PROGRAM_VERSION," starting up");
	}

	InstallSignalHandler();

	if (argc > 1) {
		free(config.config_file);
		config.config_file = AllocString(argv[1]);
	}
	
	ReadConfigFile();

	if (config.daemon) {
		ForkToBackground();
	}
	
	OpenJoystick();

	AddMissingCalibrations();

	ActionLoop();

	CloseJoystick();

	Print(stdout,PROGRAM_VERSION," quits now (should've never come around here!?)");

	return 0;
}
